### ISSUE TITLE
Log writing "savegames" in plain text for desync debugging

### ISSUE THREAD

**Issue by Nightinggale** [We-the-People-civ4col-mod/Mod#59]
When a network game desyncs, all it does is telling that some checksum differs. It's impossible to tell what went wrong from that message, meaning it can become like looking for a needle in a haystack. The cause could be anywhere.

I propose when the game is in desync mode, a button appears on the main screen. If clicked, it sends a network package to CvGame, which in turn triggers each computer will save the game _at the very same time_. Now if the format is plain text with one variable on each line, diff can be used to compare and highlight which variables contains different values. This can then give hints on what caused the desync. Maybe add that all computers make a regular savegame at the same time.

The text savegame could likely be a log output rather than an actual savegame file.

Worth considering if this can share code with #46

**Comment by ErikSandbraaten** [comment]
It could be helpful to add additional factors to the checksum like K-Mod does it:
https://github.com/karadoc/Civ4-K-Mod/commit/ba0ac7078762b422d5161005f3f2a3c2f458cc4b

This could help us detect desync issues at an earlier stage.

**Comment by Nightinggale** [comment]
I thought about modding the detection system as well. It would be obvious in order to monitor new concepts or variables, which aren't in vanilla. However detecting a desync and what to do about a desync once detected are two very different issues and we shouldn't mix them together. In other words this ticket has nothing to do with detecting the desyncs, only what to do to learn more about already detected desyncs.

The code is really stable in network games compared to other mods, but it still manage to get rare desyncs. Even worse we have no good clues to the remaining causes for desyncs, hence why I'm thinking of ways to learn more about what goes wrong rather than trying to aim at fixing the unknown.

It's also a valid argument that new desyncs can be added in the future. With proper tools they can be detected and fixed much quicker.

**Comment by Nightinggale** [comment]
I started working on this due to #190 and the fact that I have no idea why it desyncs. Status right now is that I added a button to the main screen and if I click it, it will send a network package, which makes the receivers write a txt file in the mod root dir. The local player ID is included in the file name, meaning no name clashes even if both "network" players run the same mod on the same computer. This works and I can get the same (currently mostly empty) file from both players when one player clicks the button.

I tried using fstream, but it failed to compile due to something in the fstream header. I tried to google the error and found two other people with the same error, both from 2005. According to the replies back then, it seems to be a known compiler issue and that it conflicts with something in the code, though it's not known what. The recommendation is to use C style writing. I did that and it works out of the box. Not as good for nice code, but being able to compile is a major bonus.

I will probably just add variables until I find the cause of the desync and then leave it half finished at that point. The CivEffect branch has code for loading xml files, which is useful here, as in it can convert enum (int) values to the strings from getType. Using something like that will provide a much more readable deync log, but right now my goal is a quick fix for desyncs in 2.7.1.
